app/models/assignment.rb

```diff
@@ -343,29 +377,44 @@ def all_grouping_data
     grouping_data = self
                     .groupings
                     .joins(:group)
+                    .left_outer_joins(:extension)
                     .left_outer_joins(non_rejected_student_memberships: :user)
-                    .pluck('groupings.id',
-                           'groupings.admin_approved',
-                           'groups.group_name',
-                           'users.user_name',
-                           'memberships.membership_status')
-
-    groupings = Hash.new { |h, k| h[k] = [] }
-    grouping_data.each do |gid, approved, name, user_name, status|
-      groupings[[gid, approved, name]]
-      if user_name
-        groupings[[gid, approved, name]] << [user_name, status]
-        students[user_name][:assigned] = true
+                    .pluck_to_hash('groupings.id',
+                                   'groupings.admin_approved',
+                                   'groups.group_name',
+                                   'users.user_name',
+                                   'memberships.membership_status',
+                                   'extensions.time_delta',
+                                   'extensions.apply_penalty',
+                                   'extensions.note')
+
+    members = Hash.new { |h, k| h[k] = [] }
+    grouping_data.each do |data|
+      if data['users.user_name']
+        members[data['groupings.id']] << [data['users.user_name'], data['memberships.membership_status']]
+        students[data['users.user_name']][:assigned] = true
       end
     end
-    groupings = groupings.map do |k, v|
+    ids = Set.new
+    groupings = grouping_data.map do |data|
+      next if ids.include? data['groupings.id'] # distinct on the query doesn't seem to work
+
+      ids << data['groupings.id']
+      if data['extensions.time_delta'].nil?
+        data['extensions.time_delta'] = {}
+      else
+        data['extensions.time_delta'] = Extension.to_parts ActiveSupport::Duration.parse(data['extensions.time_delta'])
+      end
       {
-        _id: k[0],
-        admin_approved: k[1],
-        group_name: k[2],
-        members: v
+        _id: data['groupings.id'],
+        admin_approved: data['groupings.admin_approved'],
+        group_name: data['groups.group_name'],
+        extension: data['extensions.time_delta'],
```

Comment:
I see what you're trying to do above (keep the week/days/hours as the special keys so you can loop through them), but I don't think the payoff is worth it.

I'd like to see a single hash here for extension that contains all of the attributes for the associated extension, including the extension's `id`. Then you can pass that around in the React components as one big blob, rather than the decomposition that's happening now. For example, the `selected_row_data` state can be just the extension data rather than the whole row.

The `id` attribute is useful because you can use it in the front-end to determine whether a new extension is being created (say, `id` is `undefined`), and to do the update/delete by (extension) `id` rather than `grouping_id`.